app/code/Magento/SalesRule/Model/ResourceModel/Rule/Collection.php

```diff
@@ -149,84 +156,125 @@ public function setValidationFilter(
-                $select->joinLeft(
-                    ['rule_coupons' => $this->getTable('salesrule_coupon')],
-                    $connection->quoteInto(
-                        'main_table.rule_id = rule_coupons.rule_id AND main_table.coupon_type != ?',
-                        \Magento\SalesRule\Model\Rule::COUPON_TYPE_NO_COUPON
-                    ),
-                    ['code']
-                );
-
-                $noCouponWhereCondition = $connection->quoteInto(
-                    'main_table.coupon_type = ? ',
-                    \Magento\SalesRule\Model\Rule::COUPON_TYPE_NO_COUPON
-                );
-
-                $autoGeneratedCouponCondition = [
-                    $connection->quoteInto(
-                        "main_table.coupon_type = ?",
-                        \Magento\SalesRule\Model\Rule::COUPON_TYPE_AUTO
-                    ),
-                    $connection->quoteInto(
-                        "rule_coupons.type = ?",
-                        \Magento\SalesRule\Api\Data\CouponInterface::TYPE_GENERATED
-                    ),
-                ];
-
-                $orWhereConditions = [
-                    "(" . implode($autoGeneratedCouponCondition, " AND ") . ")",
-                    $connection->quoteInto(
-                        '(main_table.coupon_type = ? AND main_table.use_auto_generation = 1 AND rule_coupons.type = 1)',
-                        \Magento\SalesRule\Model\Rule::COUPON_TYPE_SPECIFIC
-                    ),
-                    $connection->quoteInto(
-                        '(main_table.coupon_type = ? AND main_table.use_auto_generation = 0 AND rule_coupons.type = 0)',
-                        \Magento\SalesRule\Model\Rule::COUPON_TYPE_SPECIFIC
-                    ),
-                ];
-
-                $andWhereConditions = [
-                    $connection->quoteInto(
-                        'rule_coupons.code = ?',
-                        $couponCode
-                    ),
-                    $connection->quoteInto(
-                        '(rule_coupons.expiration_date IS NULL OR rule_coupons.expiration_date >= ?)',
-                        $this->_date->date()->format('Y-m-d')
-                    ),
-                ];
-
-                $orWhereCondition = implode(' OR ', $orWhereConditions);
-                $andWhereCondition = implode(' AND ', $andWhereConditions);
-
-                $select->where(
-                    $noCouponWhereCondition . ' OR ((' . $orWhereCondition . ') AND ' . $andWhereCondition . ')',
-                    null,
-                    Select::TYPE_CONDITION
-                );
+            $noCouponRules = $this->getNoCouponCodeSelect();
+
+            if ($couponCode) {
+                $couponRules = $this->getCouponCodeSelect($couponCode);
+
+                $allAllowedRules = $this->getConnection()->select();
+                $allAllowedRules->union([$noCouponRules, $couponRules], \Zend_Db_Select::SQL_UNION_ALL);
```

Comment:
Why do you merge sales rules without coupons ($noCouponRules) to select in case, when we filter by specific coupon?
If you will have a big amount of sales rules without a coupon code (coupon_type == Rule::COUPON_TYPE_NO_COUPON) then you will have big performance degradation.
I think we should optimize this flow

BTW, you can experiment with a big amount of sales rules using generation toolkit: https://devdocs.magento.com/guides/v2.3/config-guide/cli/config-cli-subcommands-perf-data.html (you can adjust [car_price_rule](https://github.com/magento/magento2/blob/2.3-develop/setup/performance-toolkit/profiles/ce/small.xml#L40 for generate any amount of sales rules)

Reply:
For this PR you can ignore ^^ my comment, due to it relates to current process how we apply sales rules to the cart. It can be implemented in separate PR.